pkg/snapstore/s3_snapstore.go

```diff
@@ -510,35 +518,108 @@ func (s *S3SnapStore) partUploader(wg *sync.WaitGroup, stopCh <-chan struct{}, s
 func (s *S3SnapStore) List(_ bool) (brtypes.SnapList, error) {
+	var snapList brtypes.SnapList
 	prefixTokens := strings.Split(s.prefix, "/")
 	// Last element of the tokens is backup version
 	// Consider the parent of the backup version level (Required for Backward Compatibility)
 	prefix := path.Join(strings.Join(prefixTokens[:len(prefixTokens)-1], "/"))
 
-	var snapList brtypes.SnapList
-	in := &s3.ListObjectsInput{
-		Bucket: aws.String(s.bucket),
-		Prefix: aws.String(prefix),
-	}
-	err := s.client.ListObjectsPages(in, func(page *s3.ListObjectsOutput, lastPage bool) bool {
-		for _, key := range page.Contents {
-			k := (*key.Key)[len(*page.Prefix):]
-			if strings.Contains(k, backupVersionV1) || strings.Contains(k, backupVersionV2) {
-				snap, err := ParseSnapshot(path.Join(prefix, k))
-				if err != nil {
-					// Warning
-					logrus.Warnf("Invalid snapshot found. Ignoring it: %s", k)
+	// Get the status of bucket versioning.
+	// Note: Bucket versioning will always be enabled for object lock.
+	versioningStatus, err := s.client.GetBucketVersioning(&s3.GetBucketVersioningInput{
+		Bucket: &s.bucket,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if versioningStatus.Status != nil && *versioningStatus.Status == "Enabled" {
+		// versioning is found to be enabled on given bucket.
+		logrus.Info("Bucket versioning is found to be enabled.")
+
+		isObjectLockEnabled, bucketImmutableExpiryTimeInDays, err := GetBucketImmutabilityTime(s)
+		if err != nil {
+			logrus.Warnf("unable to check object lock configuration for the bucket: %v", err)
+		}
+		if !isObjectLockEnabled {
+			logrus.Warnf("Bucket versioning is found to be enabled but object lock is not found to be enabled.")
+			logrus.Warnf("Please enable the object lock as well for the given bucket for immutability of snapshots.")
+		}
+
+		in := &s3.ListObjectVersionsInput{
+			Bucket: aws.String(s.bucket),
+			Prefix: aws.String(prefix),
+		}
+
+		if err := s.client.ListObjectVersionsPages(in, func(page *s3.ListObjectVersionsOutput, lastPage bool) bool {
+			for _, version := range page.Versions {
+				if *version.IsLatest {
+					k := (*version.Key)[len(*page.Prefix):]
+					if strings.Contains(k, backupVersionV1) || strings.Contains(k, backupVersionV2) {
+						snap, err := ParseSnapshot(path.Join(prefix, k))
+						if err != nil {
+							// Warning
+							logrus.Warnf("Invalid snapshot found. Ignoring it: %s", k)
+						} else {
+							// capture the versionID of snapshot and immutability expiry time of snapshot.
+							snap.VersionID = *version.VersionId
+							if bucketImmutableExpiryTimeInDays != nil {
+								// To get S3 object's "RetainUntilDate" or "ImmutabilityExpiryTime", backup-restore need to make an API call for each snapshots.
+								// To avoid API calls for each snapshots, backup-restore is calculating the "ImmutabilityExpiryTime" using bucket retention period.
```

Comment:
```suggestion
								// To get S3 object's "RetainUntilDate" or "ImmutabilityExpiryTime", backup-restore need to make an API call for each snapshot.
								// To avoid API calls for each snapshot, backup-restore is calculating the "ImmutabilityExpiryTime" using bucket retention period.
```